Alignment/HIPAlignmentAlgorithm/plugins/HIPTwoBodyDecayAnalyzer.cc

```diff
@@ -0,0 +1,385 @@
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  TTree* tree;
+  std::vector<std::pair<TString, Float_t*>> floatBranches;
+  std::vector<std::pair<TString, Int_t*>> intBranches;
+  std::vector<std::pair<TString, Short_t*>> shortBranches;
+
+private:
+
+  virtual void beginJob() override;
+  virtual void analyze(const edm::Event&, const edm::EventSetup&) override;
+  virtual void endJob() override;
+
+  void bookAllBranches();
+  bool bookBranch(TString bname, TString btype);
+  TString searchArray(TString branchname, int& position);
+  template<typename varType> void setVal(TString bname, varType value){
+    int varposition=-1;
+    TString varbranchtype = searchArray(bname, varposition);
+    if (varposition==-1) std::cerr << "HIPTwoBodyDecayAnalyzer::setVal -> Could not find the branch called " << bname << "!" << std::endl;
+    else if (varbranchtype=="s") *(shortBranches.at(varposition).second)=value;
+    else if (varbranchtype=="i") *(intBranches.at(varposition).second)=value;
+    else if (varbranchtype=="f") *(floatBranches.at(varposition).second)=value;
+    else std::cerr << "HIPTwoBodyDecayAnalyzer::setVal -> Could not find the type " << varbranchtype << " for branch called " << bname << "!" << std::endl;
+  }
+  void cleanBranches();
+  void initializeBranches();
+  bool actuateBranches();
+
+  void analyzeTrackCollection(
+    TString strTrackType,
+    edm::ESHandle<TransientTrackBuilder>& theTTBuilder,
+    edm::Handle<reco::TrackCollection>& hTrackColl,
+    bool verbose=false
+    );
+  reco::Vertex fitDimuonVertex(
+    edm::ESHandle<TransientTrackBuilder>& theTTBuilder,
+    edm::Handle<reco::TrackCollection>& hTrackColl,
+    bool& fitOk
+    );
+
+};
+
+HIPTwoBodyDecayAnalyzer::HIPTwoBodyDecayAnalyzer(const edm::ParameterSet& iConfig){
+  alcareco_trackCollToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("alcarecotracks"));
+  refit1_trackCollToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("refit1tracks"));
+  ctf_trackCollToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("refit2tracks"));
+  final_trackCollToken_ = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("finaltracks"));
+
+  edm::Service<TFileService> fs;
+
+  tree = fs->make<TTree>("TestTree", "");
+  bookAllBranches();
+}
+
+
+TString HIPTwoBodyDecayAnalyzer::searchArray(TString branchname, int& position){
+  for (unsigned short el=0; el<shortBranches.size(); el++){
+    if (branchname==shortBranches.at(el).first){
+      position = el;
+      return "s";
+    }
+  }
+  for (unsigned int el=0; el<intBranches.size(); el++){
+    if (branchname==intBranches.at(el).first){
+      position = el;
+      return "i";
+    }
+  }
+  for (unsigned int el=0; el<floatBranches.size(); el++){
+    if (branchname==floatBranches.at(el).first){
+      position = el;
+      return "f";
+    }
+  }
+  return "unknown";

```

Comment:
I agree that the current functionality could be improved a lot. In fact I don't think I need any enums here as long as I have template functions and instantiation done more properly. However, I also think that maybe I can make a separate PR for this instead of delaying this one. Overall we made a lot of improvements that we would also like different people in the Tracker Alignment group to be able to use, but we cannot really put these new functionalities into use coherently without merging the PR. May I ask if you would agree that we open another PR for these changes instead of making this one wait?